hypothesis-python/src/hypothesis/internal/conjecture/shrinking/dfas.py

```diff
@@ -0,0 +1,336 @@
+            # we learn a DFA that could match a string longer
+            # than len(v_core) we fix it by finding the first
+            # string longer than v_core and learning that as
+            # a correction.
+            x = next(learner.dfa.all_matching_strings(min_length=len(v_core) + 1))
+            assert not is_valid_core(x)
+            learner.learn(x)
+            assert not learner.dfa.matches(x)
+            assert learner.generation > prev
+        else:
+            # We mostly care about getting the right answer on the
+            # minimal test case, but because we're doing this offline
+            # anyway we might as well spend a little more time trying
+            # small examples to make sure the learner gets them right.
+            for x in islice(learner.dfa.all_matching_strings(), 100):
+                if not is_valid_core(x):
+                    learner.learn(x)
+                    assert learner.generation != prev
+                    break
+
+    # We've now successfully learned a DFA that works for shrinking
+    # our failed normalisation further. Canonicalise it into a concrete
+    # DFA so we can save it for later.
+    new_dfa = learner.dfa.canonicalise()
+
+    assert math.isfinite(new_dfa.max_length(new_dfa.start))
+
+    shrinker = runner.new_shrinker(runner.cached_test_function(v), predicate)
+
+    assert (len(prefix), len(v) - len(suffix)) in shrinker.matching_regions(new_dfa)
+
+    name = "tmp-dfa-" + repr(new_dfa)
+
+    shrinker.extra_dfas[name] = new_dfa
+
+    shrinker.fixate_shrink_passes([dfa_replacement(name)])
+
+    assert sort_key(shrinker.buffer) < sort_key(v)
+
+    return new_dfa
+
+
+def fully_shrink(runner, test_case, predicate):
+    if not isinstance(test_case, ConjectureResult):
+        test_case = runner.cached_test_function(test_case)
+    while True:
+        shrunk = runner.shrink(test_case, predicate)
+        if shrunk.buffer == test_case.buffer:
+            break
+        test_case = shrunk
+    return test_case
+
+
+def normalize(
+    base_name,
+    test_function,
+    *,
+    required_successes=100,
+    allowed_to_update=False,
+    max_dfas=10
+):
+    """Attempt to ensure that this test function successfully normalizes - i.e.
+    whenever it declares a test case to be interesting, we are able
+    to shrink that to the same interesting test case (which logically should
+    be the shortlex minimal interesting test case, though we may not be able
+    to detect if it is).
+
+    Will run until we have seen ``required_successes`` many interesting test
+    cases in a row normalize to the same value.
+
+    If ``allowed_to_update`` is True, whenever we fail to normalize we will
+    learn a new DFA-based shrink pass that allows us to make progress. Any
+    learned DFAs will be written back into this file at the end of this
+    function. If ``allowed_to_update`` is False, this will raise an error
+    as soon as it encounters a failure to normalize.
```

Comment:
“back into this file” is slightly misleading, since the learned DFAs end up written to a *different* file.